File: content/en/04-llms/llms/_exercises/51_rag/51_rag.R
#+ setup
library(ragnar)

# Step 1: Read, chunk and create embeddings for "R for Data Science" ----------

#' This example is based on https://ragnar.tidyverse.org/#usage.
#'
#' The first step is to crawl the R for Data Science website to find all the
#' pages we'll need to read in.
#'
#' Then, we create a new ragnar document store that will use the LM Studio
#' (https://lmstudio.ai/) `text-embedding-nomic-embed-text-v2-moe` model to
#' create embeddings for each chunk of text. LM Studio exposes an
#' OpenAI-compatible API at http://localhost:1234/v1, so make sure it is
#' running with that model loaded:
#' https://huggingface.co/nomic-ai/nomic-embed-text-v2-moe-GGUF
#'
#' Finally, we read each page as markdown, use `markdown_chunk()` to split that
#' markdown into reasonably-sized chunks, finally inserting each chunk into the
#' vector store. That insertion step automatically sends the chunk text to
#' LM Studio to create the embedding, and ragnar stores the embedding alongside
#' the original text of the chunk.

#+ create-store

base_url <- "https://r4ds.hadley.nz"
pages <- ragnar_find_links(base_url, children_only = TRUE)

store_location <- here::here("_exercises/51_rag/r4ds.ragnar.duckdb")

store <- ragnar_store_create(
  store_location,
  title = "R for Data Science",
  # Need to start over? Set `overwrite = TRUE`.
  # overwrite = TRUE,
  embed = \(x) {
    embed_lm_studio(x, model = "text-embedding-nomic-embed-text-v2-moe")
  }
)

cli::cli_progress_bar(total = length(pages))
for (page in pages) {
  cli::cli_progress_update(status = page)

  chunks <- page |>
    read_as_markdown() |>
    # The next step breaks the markdown into chunks. This is where you have the
    # most control over what content is grouped together for embedding and later
    # retrieval. Feel free to experiment with settings in `?markdown_chunk()`.
    markdown_chunk()

  ragnar_store_insert(store, chunks)
}
cli::cli_progress_done()

ragnar_store_build_index(store)

# Step 2: Inspect your document store -----------------------------------------

#' Now that we have the vector store, what chunks are surfaced when we ask a
#' question? To do that, we'll use the ragnar store inspector app and an
#' example question.
#'
# Here's a question someone might ask an LLM. Copy the task markdown to use in
# the ragnar store inspector app.

#+ inspect-store
task <- r"--(
Could someone help me filter one data frame by matching values in another?

I’ve got two data frames with a common column `code.` I want to keep rows in `data1` where `code` exists in `data2$code`. I tried using `filter()` but got no rows back.

Here’s a minimal example:

```r
library(dplyr)

data1 <- data.frame(
    closed_price = c(49900L, 46900L, 46500L),
    opened_price = c(51000L, 49500L, 47500L),
    adjust_closed_price = c(12951L, 12173L, 12069L),
    stock = as.factor(c("AAA", "AAA", "AAC")),
    date3 = as.factor(c("2010-07-15", "2011-07-19", "2011-07-23")),
    code = as.factor(c("AAA2010", "AAA2011", "AAC2011"))
)

data2 <- data.frame(
    code = as.factor(c("AAA2010", "AAC2011")),
    ticker = as.factor(c("AAA", "AAM"))
)
```

What I tried:

```r
price_code <- data1 %>% filter(code %in% data2)
```

This returns zero rows. What’s the simplest way to do this?
)--"

ragnar_store_inspect(store)


# Step 3: Use document store in a chatbot --------------------------------------

#' Finally, ragnar provides a special tool that attaches to an ellmer chat
#' client and lets the model retrieve relevant chunks from the vector store on
#' demand. Run the code below to launch a chatbot backed by all the knowledge in
#' the R for Data Science book. Paste the task markdown from above into the chat
#' and see how the chatbot uses the retrieved chunks to improve its answer, or
#' ask it your own questions about R for Data Science.

#+ chatbot

library(ellmer)

chat <- chat_posit(
  model = "zai-org/GLM-5.3-Flash",
  system_prompt = r"--(
You are an expert R programmer and mentor. You are concise.

Before responding, retrieve relevant material from the knowledge store. Quote or
paraphrase passages, clearly marking your own words versus the source. Provide a
working link for every source you cite.
  )--"
)

# Attach the retrieval tool to the chat client. You can choose how many chunks
# or documents are retrieved each time the model uses the tool.
ragnar_register_tool_retrieve(chat, store, top_k = 10)

live_console(chat)

File: content/en/04-llms/llms/_solutions/51_rag/51_rag.py
# %%
import chatlas

# %% [markdown]
# [raghilda](https://posit-dev.github.io/raghilda/) is the Python counterpart
# to `ragnar`: it turns a pile of documents into a searchable knowledge store.
# We'll point it at the markdown files from the
# [Polars Cookbook](https://github.com/escobar-west/polars-cookbook), which
# live in `data/polars-cookbook`.
#
# The pipeline has three steps. `read_as_markdown()` reads each file as
# markdown. `MarkdownChunker()` splits each file into chunks, and this is where
# you have the most control over what content is grouped together for embedding
# and later retrieval. Finally, `store.upsert()` inserts each chunk into a
# DuckDB database, sending the chunk text to the embedding provider on the way
# in. raghilda stores the embedding alongside the original text of the chunk.
#
# Embeddings are served by LM Studio (https://lmstudio.ai/), which exposes an
# OpenAI-compatible API on localhost. Make sure LM Studio is running with the
# text-embedding-nomic-embed-text-v2-moe model loaded:
# https://huggingface.co/nomic-ai/nomic-embed-text-v2-moe-GGUF
# %%
from pyhere import here
from raghilda.chunker import MarkdownChunker
from raghilda.embedding import EmbeddingOpenAI
from raghilda.read import read_as_markdown
from raghilda.store import DuckDBStore

embed = EmbeddingOpenAI(
    model="text-embedding-nomic-embed-text-v2-moe",
    base_url="http://localhost:1234/v1",
    api_key="lm-studio",
)

store_location = here("_solutions/51_rag/polars_cookbook.raghilda.duckdb")

store = DuckDBStore.create(
    store_location,
    title="Polars Cookbook",
    # Need to start over? Set `overwrite = True`.
    # overwrite = True,
    embed=embed,
)

chunker = MarkdownChunker()

cookbook_dir = here("data/polars-cookbook")
for file in sorted(cookbook_dir.glob("*.md")):
    document = read_as_markdown(str(file))
    chunked_document = chunker.chunk(document)
    store.upsert(chunked_document)

store.build_index()

# %% [markdown]
# Now that the store is on disk, which chunks surface for a given question?
# Connect to the store with `DuckDBStore.connect()`. raghilda remembers the
# embedding provider used to build the store, so you don't need to configure
# it again.
#
# Next, write `retrieve_polars_knowledge()`, the function the LLM will call as
# a tool. The docstring is sent to the model as the tool description, so write
# it for the LLM: be specific about when the tool should be used and what the
# `query` should contain. Tool results are sent to the model as text, so we
# return JSON. It preserves structure and lets the model parse the results
# without any extra work. We also include the `origin` of each chunk, the file
# it came from, so the model can tell the user where it found the material.

# %%
import json

store = DuckDBStore.connect(store_location, read_only=True)


def retrieve_polars_knowledge(query: str, num_results: int = 5) -> str:
    """
    Search the Polars Cookbook for relevant content.

    Use this tool when the user asks about polars syntax, such as selecting
    columns, filtering rows, grouping, or aggregating data.

    Parameters
    ----------
    query : str
        A description of what to look for.
    num_results : int
        The number of relevant passages to return (default of 5).
    """
    chunks = store.retrieve(query, top_k=num_results, deoverlap=True)
    return json.dumps(
        [{"text": chunk.text, "source": chunk.origin} for chunk in chunks]
    )


# %% [markdown]
# This implementation retrieves the 5 most relevant chunks for the query. You
# can change the number of results with the `num_results` parameter. There's no
# magic number, but you may want to increase it if the retrieved content is
# too sparse or not relevant enough.
#
# Let's try it out now with a task:

# %%
task = """
How do I find all rows in a DataFrame which have the max value for count column, after grouping by ['Sp','Mt'] columns?

Example 1: the following DataFrame, which I group by ['Sp','Mt']:

```
Sp Mt Value count
0 MM1 S1 a 2
1 MM1 S1 n **3**
2 MM1 S3 cb **5**
3 MM2 S3 mk **8**
4 MM2 S4 bg **5**
5 MM2 S4 dgd 1
6 MM4 S2 rd 2
7 MM4 S2 cb 2
8 MM4 S2 uyi **7**
```

Expected output: get the result rows whose count is max in each group, like:

```
1 MM1 S1 n **3**
2 MM1 S3 cb **5**
3 MM2 S3 mk **8**
4 MM2 S4 bg **5**
8 MM4 S2 uyi **7**
```
"""

retrieve_polars_knowledge(task)

# %% [markdown]
# Finally, register the retrieval tool with a chatlas chatbot. Copy the task
# from the previous block and paste it into the chatbot, or ask it your own
# questions about the Polars Cookbook.

# %%
chat = chatlas.ChatPosit(
    model="zai-org/GLM-5.3-Flash",
    system_prompt="""
You are an expert Python programmer and mentor. You are concise.

Before responding, retrieve relevant material from the knowledge store. Quote or
paraphrase passages, clearly marking your own words versus the source, and tell
the user which cookbook chapter each passage came from.
""",
)

chat.register_tool(retrieve_polars_knowledge)

chat.app()
